gcputil/credentials.go

```diff
@@ -50,6 +53,127 @@ type GcpCredentials struct {
 	ProjectId    string `json:"project_id" structs:"project_id" mapstructure:"project_id"`
 }
 
+type ExternalAccountCredential struct {
+	// External Account fields
+	Audience              string   `json:"audience"`
+	ServiceAccountEmail   string   `json:"service_account_email"`
+	SubjectTokenType      string   `json:"subject_token_type"`
+	TokenURL              string   `json:"token_url"`
+	Scopes                []string `json:"scopes"`
+	WorkloadIdentityToken string   `json:"workload_identity_token"`
+}
+
+func (c *ExternalAccountCredential) TokenSource(ctx context.Context) (oauth2.TokenSource, error) {
+	ts := tokenSource{
+		ctx:    ctx,
+		config: c,
+	}
+	return oauth2.ReuseTokenSource(nil, ts), nil
+}
+
+func (c *ExternalAccountCredential) GetCredentials(ctx context.Context) (*google.Credentials, error) {
+	ts, err := c.TokenSource(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &google.Credentials{
+		TokenSource: ts,
+	}, nil
+}
+
+// tokenSource is the source that handles external credentials. It is used to retrieve Tokens.
+type tokenSource struct {
+	ctx    context.Context
+	config *ExternalAccountCredential
+}
+
+// Token allows tokenSource to conform to the oauth2.TokenSource interface.
+func (ts tokenSource) Token() (*oauth2.Token, error) {
```

Comment:
It's hard to say exactly without seeing the consumer of this method. But we may want to add a lock since the Token() implementation [must be thread safe](https://github.com/golang/oauth2/blob/4b7f0bdbc7c30d05383d2597847300fbb870608b/oauth2.go#L70).

Reply:
Ah good call, will test adding mutex locks here as well

Reply:
Now that I have seen the usage of this method I think we do want a Read lock on the tokenSource config.